compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala

```diff
@@ -109,42 +129,53 @@ abstract class PrepJSInterop extends plugins.PluginComponent
     /** Whether to check and prepare exports. */
     private def shouldPrepareExports = !forScaladoc
 
-    private def jsAnyClassOnly = !inJSAnyCls && allowJSAny
-    private def allowImplDef   = !inJSAnyCls && !inJSAnyMod
-    private def allowJSAny     = !inScalaCls
-    private def inJSAny        = inJSAnyMod || inJSAnyCls
+    /** Whether we can define JS native entities in the current scope. */
+    private def allowJSNative =
+      noEnclosingOwner is OwnerKind.AnyClass
+
+    /** Whether we can define JS native classes in the current scope. */
+    private def allowJSNativeClass =
+      noEnclosingOwner is OwnerKind.AnyClass
+
+    /** Whether we can define any class/trait/method in the current scope.
+     *
+     *  This is overridden by [[allowJSNativeClass]] if the latter is true.
+     */
+    private def allowImplDef = noEnclosingOwner is OwnerKind.JSNative
 
     /** DefDefs in class templates that export methods to JavaScript */
     private val exporters = mutable.Map.empty[Symbol, mutable.ListBuffer[Tree]]
 
     override def transform(tree: Tree): Tree = postTransform { tree match {
       // Catch special case of ClassDef in ModuleDef
-      case cldef: ClassDef if jsAnyClassOnly && isJSAny(cldef) =>
+      case cldef: ClassDef if allowJSNativeClass && isJSAny(cldef) =>
         transformJSAny(cldef)
 
       // Catch forbidden implDefs
       case idef: ImplDef if !allowImplDef =>
-        reporter.error(idef.pos, "Traits, classes and objects extending " +
-            "js.Any may not have inner traits, classes or objects")
+        reporter.error(idef.pos, "Native JS traits, classes and objects " +
+            "may not have inner traits, classes or objects")
         super.transform(tree)
 
       // Handle js.Anys
       case idef: ImplDef if isJSAny(idef) =>
         transformJSAny(idef)
 
+      // @ScalaJSDefined is only valid on a js.Any
+      case idef: ImplDef if idef.symbol.hasAnnotation(ScalaJSDefinedAnnotation) =>
+        reporter.error(idef.pos,
+            "@ScalaJSDefined is only allowed on classes extending js.Any")
+        super.transform(tree)
+
       // Catch the definition of scala.Enumeration itself
       case cldef: ClassDef if cldef.symbol == ScalaEnumClass =>
-        enterEnumImpl { super.transform(cldef) }
+        enterOwner(OwnerKind.EnumImpl) { super.transform(cldef) }
 
       // Catch Scala Enumerations to transform calls to scala.Enumeration.Value
       case cldef: ClassDef if isScalaEnum(cldef) =>
-        enterScalaCls {
-          enterScalaEnum {
-            super.transform(cldef)
-          }
-        }
+        enterOwner(OwnerKind.EnumClass) { super.transform(cldef) }
       case idef: ImplDef if isScalaEnum(idef) =>
-        enterScalaEnum { super.transform(idef) }
+        enterOwner(OwnerKind.EnumMod) { super.transform(idef) }
```

Comment:
Consider pulling these two cases in a single case (and have a short `if` to separate `EnumMod` from `EnumClass`). Or at least change the lower match from `ImplDef` to `ModuleDef`.